mist-common/src/main/java/edu/snu/mist/common/operators/OneStreamStateHandlerOperator.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 2018 Seoul National University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package edu.snu.mist.common.operators;
+
+import edu.snu.mist.common.MistEvent;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+public abstract class OneStreamStateHandlerOperator extends OneStreamOperator implements StateHandler {
+
+  /**
+   * The latest Checkpoint Timestamp.
+   */
+  protected long latestTimestampBeforeCheckpoint;
+
+  /**
+   * The recovered latest Checkpoint Timestamp.
+   */
+  protected long recoveredCheckpointTimestamp;
+
+  /**
+   * The map of states for checkpointing.
+   * The key is the timestamp of the last event before a checkpoint event,
+   * and the value is the state of this operator at that timestamp.
+   */
+  protected TreeMap<Long, Map<String, Object>> checkpointMap;
+
+  protected OneStreamStateHandlerOperator() {
+    this.latestTimestampBeforeCheckpoint = 0L;
+    this.recoveredCheckpointTimestamp = 0L;
+    this.checkpointMap = new TreeMap<>();
```

Comment:
Doesn't it have concurrency issue? While the checkpoint thread tries to read the map, the event processor can write the state to the map. Or `removeOldStates` could be performed concurrently with reader threads.

Reply:
I'll use ConcurrentSkipListMap instead of TreeMap.